<!DOCTYPE html>
<html>
<head>
  <meta charset="utf-8">
  <meta name="generator" content="pandoc">
  <meta name="author" content="Peter Ralph">
  <title>Tree sequences</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="reveal.js/css/reveal.css">
  <style type="text/css">
      code{white-space: pre-wrap;}
      span.smallcaps{font-variant: small-caps;}
      span.underline{text-decoration: underline;}
      div.column{display: inline-block; vertical-align: top; width: 50%;}
  </style>
  <link rel="stylesheet" href="reveal.js/css/theme/simple.css" id="theme">
  <!-- Printing and PDF exports -->
  <script>
    var link = document.createElement( 'link' );
    link.rel = 'stylesheet';
    link.type = 'text/css';
    link.href = window.location.search.match( /print-pdf/gi ) ? 'reveal.js/css/print/pdf.css' : 'reveal.js/css/print/paper.css';
    document.getElementsByTagName( 'head' )[0].appendChild( link );
  </script>
  <!--[if lt IE 9]>
  <script src="reveal.js/lib/js/html5shiv.js"></script>
  <![endif]-->
  <style type="text/css">
  
  .reveal { font-size: 30px; }
  
  .reveal h1 { font-size: 1.5em; } 
  
  .reveal h2 { font-size: 1.2em; } 
  
  .reveal .slides { text-align: left; }
  
  .reveal .slides figure { text-align: center; }
  
  .centered p { text-align: center; }
  
  .caption p { font-size: 75%; }
  
  .greyed { color: grey; }
  
  div.floatright { float: right; text-align: right; }
  
  .reveal figcaption { display: none; }
  
  .reveal img { border: none; }
  
  .reveal section img { border: none; }
  
  .MathJax_Display { margin: 0; }
  
  img.finger { vertical-align: text-top; }
  
  </style>
  <div style="display: none">
  \[
  \newcommand{\E}{\mathbb{E}}
  \renewcommand{\P}{\mathbb{P}}
  \DeclareMathOperator{\var}{var}
  \]
  </div>
</head>
<body>
  <div class="reveal">
    <div class="slides">

<section id="title-slide">
  <h1 class="title">Tree sequences</h1>
  <p class="subtitle">progress and challenges in genomes and genealogies</p>
  <p class="author">Peter Ralph</p>
  <p class="date">Northwest Probability // October 2019</p>
</section>

<section><section id="genomes-and-genealogies" class="title-slide slide level1"><h1>Genomes and genealogies</h1></section><section id="genomes" class="slide level2">
<h2>Genomes</h2>
<ul>
<li class="fragment">are very big (<span class="math inline">\(10^7\)</span>–<span class="math inline">\(10^12\)</span> nucleotides)</li>
<li class="fragment">encode the basic <em>mechanisms</em> of life</li>
<li class="fragment">reflect past <em>history</em> and <em>process</em></li>
</ul>
</section><section id="section" class="slide level2">
<h2></h2>
<p>Your two copies of the genome were inherited, noisily, from your two parents,</p>
<div class="fragment">
<p>and from your four grandparents,</p>
</div>
<div class="fragment">
<p>and your eight great-grandparents,</p>
</div>
<div class="fragment">
<p>see where this is going?</p>
</div>
</section><section id="genealogical-ancestors" class="slide level2">
<h2>Genealogical ancestors</h2>
<p>In a well-mixed population of size <span class="math inline">\(N\)</span>,</p>
<ul>
<li><p>A most recent common ancestor of <em>everyone</em> lived at <span class="math inline">\(\approx \log 2 N\)</span> generations ago.</p></li>
<li><p>For humans: <span class="math inline">\(\approx 500\)</span> years; but including geography and history, <span class="math inline">\(\approx\)</span> 2–3 thousand years (Chang ’99; Rohde, Olson &amp; Chang ’04)</p></li>
<li><p><em>Everyone</em> alive <span class="math inline">\(\approx 1.77 \log_2 n\)</span> generations ago is either an ancestor of everyone or has left behind no descendants. (Chang ’99)</p></li>
</ul>
</section><section id="meiosis-recombination-sex" class="slide level2">
<h2>Meiosis &amp; Recombination (Sex)</h2>
<div class="columns">
<div class="column" style="width:50%;">
<figure>
<img data-src="figs/recombination-cartoon-crop.png" alt="recombination" /><figcaption>recombination</figcaption>
</figure>
</div><div class="column" style="width:50%;">
<ul>
<li><p>You have two copies of each chromosome, one from each parent.</p></li>
<li><p>When you make a gamete, the copies <em>recombine</em>.</p></li>
<li><p>Model: switch parents at locations of a Poisson process.</p></li>
</ul>
</div>
</div>
</section><section id="you-and-your-greatk-1-grand-parents" class="slide level2">
<h2>You and your great<span class="math inline">\({}^{k-1}\)</span>-grand parents</h2>
<p>The number of <span class="math inline">\(k^\text{th}\)</span> <em>genetic</em> ancestors grows <em>linearly</em> with <span class="math inline">\(k\)</span>.</p>
<p>fig from gcbias</p>
<!-- the ARG -->
</section></section>
<section><section id="tracing-genetic-ancestry" class="title-slide slide level1"><h1>Tracing genetic ancestry</h1></section><section id="the-coalescent-with-recombination" class="slide level2">
<h2>The coalescent with recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>In a <em>randomly mating</em> population of size <span class="math inline">\(N\)</span>, ancestry of some chromosomes follows a <em>fragmentation-coalescence process</em>:</p>
<ul>
<li>labeled ancestral blocks assort into <em>ancestors</em></li>
<li><span class="greyed"> each pair of ancestors <em>coalesces</em> at rate 1 </span></li>
<li><span class="greyed"> ancestors <em>split</em> at a uniform location at rate 1 </span></li>
</ul>
<div class="greyed">
<p>Result: the <em>Ancestral Recombination Graph</em>, or ARG. (Griffiths &amp; Marjoram)</p>
</div>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/frag_coal/fragmentation_coalescence.1.png" style="width:80.0%" /></p>
</div>
</div>
</div>
</section><section id="the-coalescent-with-recombination-1" class="slide level2">
<h2>The coalescent with recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>In a <em>randomly mating</em> population of size <span class="math inline">\(N\)</span>, ancestry of some chromosomes follows a <em>fragmentation-coalescence process</em>:</p>
<ul>
<li>labeled ancestral blocks assort into <em>ancestors</em></li>
<li>each pair of ancestors <em>coalesces</em> at rate 1</li>
<li><span class="greyed"> ancestors <em>split</em> at a uniform location at rate 1 </span></li>
</ul>
<div class="greyed">
<p>Result: the <em>Ancestral Recombination Graph</em>, or ARG. (Griffiths &amp; Marjoram)</p>
</div>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/frag_coal/fragmentation_coalescence.2.png" style="width:80.0%" /></p>
</div>
</div>
</div>
</section><section id="the-coalescent-with-recombination-2" class="slide level2">
<h2>The coalescent with recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>In a <em>randomly mating</em> population of size <span class="math inline">\(N\)</span>, ancestry of some chromosomes follows a <em>fragmentation-coalescence process</em>:</p>
<ul>
<li>labeled ancestral blocks assort into <em>ancestors</em></li>
<li>each pair of ancestors <em>coalesces</em> at rate 1</li>
<li>ancestors <em>split</em> at a uniform location at rate 1</li>
</ul>
<div class="greyed">
<p>Result: the <em>Ancestral Recombination Graph</em>, or ARG. (Griffiths &amp; Marjoram)</p>
</div>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/frag_coal/fragmentation_coalescence.3.png" style="width:80.0%" /></p>
</div>
</div>
</div>
</section><section id="the-coalescent-with-recombination-3" class="slide level2">
<h2>The coalescent with recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>In a <em>randomly mating</em> population of size <span class="math inline">\(N\)</span>, ancestry of some chromosomes follows a <em>fragmentation-coalescence process</em>:</p>
<ul>
<li>labeled ancestral blocks assort into <em>ancestors</em></li>
<li>each pair of ancestors <em>coalesces</em> at rate 1</li>
<li>ancestors <em>split</em> at a uniform location at rate 1</li>
</ul>
<div class="greyed">
<p>Result: the <em>Ancestral Recombination Graph</em>, or ARG. (Griffiths &amp; Marjoram)</p>
</div>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/frag_coal/fragmentation_coalescence.4.png" style="width:80.0%" /></p>
</div>
</div>
</div>
</section><section id="the-coalescent-with-recombination-4" class="slide level2">
<h2>The coalescent with recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>In a <em>randomly mating</em> population of size <span class="math inline">\(N\)</span>, ancestry of some chromosomes follows a <em>fragmentation-coalescence process</em>:</p>
<ul>
<li>labeled ancestral blocks assort into <em>ancestors</em></li>
<li>each pair of ancestors <em>coalesces</em> at rate 1</li>
<li>ancestors <em>split</em> at a uniform location at rate 1</li>
</ul>
<div class="greyed">
<p>Result: the <em>Ancestral Recombination Graph</em>, or ARG. (Griffiths &amp; Marjoram)</p>
</div>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/frag_coal/fragmentation_coalescence.5.png" style="width:80.0%" /></p>
</div>
</div>
</div>
</section><section id="the-coalescent-with-recombination-5" class="slide level2">
<h2>The coalescent with recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>In a <em>randomly mating</em> population of size <span class="math inline">\(N\)</span>, ancestry of some chromosomes follows a <em>fragmentation-coalescence process</em>:</p>
<ul>
<li>labeled ancestral blocks assort into <em>ancestors</em></li>
<li>each pair of ancestors <em>coalesces</em> at rate 1</li>
<li>ancestors <em>split</em> at a uniform location at rate 1</li>
</ul>
<div class="greyed">
<p>Result: the <em>Ancestral Recombination Graph</em>, or ARG. (Griffiths &amp; Marjoram)</p>
</div>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/frag_coal/fragmentation_coalescence.6.png" style="width:80.0%" /></p>
</div>
</div>
</div>
</section><section id="the-coalescent-with-recombination-6" class="slide level2">
<h2>The coalescent with recombination</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>In a <em>randomly mating</em> population of size <span class="math inline">\(N\)</span>, ancestry of some chromosomes follows a <em>fragmentation-coalescence process</em>:</p>
<ul>
<li>labeled ancestral blocks assort into <em>ancestors</em></li>
<li>each pair of ancestors <em>coalesces</em> at rate 1</li>
<li>ancestors <em>split</em> at a uniform location at rate 1</li>
</ul>
<p>Result: the <em>Ancestral Recombination Graph</em>, or ARG. (Griffiths &amp; Marjoram)</p>
</div><div class="column" style="width:50%;">
<div class="floatright">
<p><img data-src="figs/frag_coal/fragmentation_coalescence.6.png" style="width:80.0%" /></p>
</div>
</div>
</div>
</section><section id="simulation" class="slide level2">
<h2>Simulation</h2>
<p><a href="https://github.com/jeromekelleher/msprime"><code>msprime</code></a>: whole genomes with <span class="math inline">\(N \sim 10^4\)</span> in <span class="math inline">\(O(\text{hours})\)</span></p>
<p>by <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher, Etheridge, &amp; McVean 2016</a></p>
<div class="floatright">
<figure>
<img data-src="figs/jerome.jpeg" alt="jerome kelleher" style="width:50.0%" /><figcaption>jerome kelleher</figcaption>
</figure>
<div class="caption">
<p>jerome kelleher</p>
</div>
</div>
</section></section>
<section><section id="forwards-simulation" class="title-slide slide level1"><h1>Forwards simulation</h1></section><section id="section-1" class="slide level2">
<h2></h2>
<p>Coalescent simulations are <em>much faster</em> than forwards-time, individual-based simulations</p>
<div class="fragment">
<p>because they don’t have to keep track of <em>everyone</em>, only the ancestors of your sample.</p>
</div>
<div class="fragment">
<p><strong>But:</strong> coalescent theory relies on a <em>duality</em> that is broken by natural selection or real geography.</p>
</div>
</section><section class="slide level2">

<p>To study evolution or understand populations across real geography, we need individual-based, forwards-time simulations.</p>
<div class="fragment">
<p>Then every individual needs to carry around her genotype (somehow). Even at neutral sites!</p>
</div>
<div class="fragment">
<p><strong>Bummer.</strong></p>
</div>
<div class="fragment">
<p><em>But wait…</em></p>
</div>
</section><section id="section-2" class="slide level2">
<h2></h2>
<p>If we <em>record the tree sequence</em> that relates everyone to everyone else,</p>
<p>after the simulation is over we can put neutral mutations down on the trees.</p>
<div class="fragment">
<p>Since neutral mutations don’t affect demography,</p>
<p>this is <em>equivalent</em> to having kept track of them throughout.</p>
</div>
</section><section class="slide level2">

<p>This means recording the entire genetic history of <strong>everyone</strong> in the population, <strong>ever</strong>.</p>
<div class="fragment">
<p>It is <em>not</em> clear this is a good idea.</p>
<!-- explain tree sequences and why they are so efficient (5min) -->
</div>
</section></section>
<section><section id="the-tree-sequence" class="title-slide slide level1"><h1>The tree sequence</h1></section><section id="history-is-a-sequence-of-trees" class="slide level2">
<h2>History is a sequence of trees</h2>
<p>For a set of sampled chromosomes, at each position along the genome there is a genealogical tree that says how they are related.</p>
<div class="fragment">
<figure>
<img data-src="figs/ts_ex/sim_ts.anim.gif" alt="Trees along a chromosome" style="width:100.0%" /><figcaption>Trees along a chromosome</figcaption>
</figure>
</div>
</section><section class="slide level2">

<p>The <strong>tree sequence</strong> is a way to describe this, er, sequence of trees.</p>
</section><section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_genotypes.png" alt="genotypes" /><figcaption>genotypes</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_genotype_and_tree.png" alt="genotypes and a tree" /><figcaption>genotypes and a tree</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="figs/ts_ex/tree_sequence_next_genotype_and_tree.png" alt="genotypes and the next tree" /><figcaption>genotypes and the next tree</figcaption>
</figure>
</section><section class="slide level2">

<p><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher, Etheridge, and McVean</a> introduced the <strong>tree sequence</strong> data structure for a fast coalescent simulator, <a href="https://github.com/jeromekelleher/msprime">msprime</a>.</p>
<ul>
<li><p>stores sequence <em>and</em> genealogical data <strong>very</strong> efficiently</p></li>
<li><p>tree-based sequence storage closely related to haplotype-matching compression</p></li>
<li><p>python/C <a href="https://github.com/tskit-dev/tskit"><code>tskit</code></a> tools</p></li>
</ul>
<div class="floatright">
<figure>
<img data-src="figs/jerome.jpeg" alt="jerome kelleher" style="width:50.0%" /><figcaption>jerome kelleher</figcaption>
</figure>
<div class="caption">
<p>jerome kelleher</p>
</div>
</div>
</section><section id="file-sizes" class="slide level2">
<h2>File sizes</h2>
<div class="centered">
<figure>
<img data-src="figs/tsinfer_sizes.png" alt="file sizes" style="width:90.0%" /><figcaption>file sizes</figcaption>
</figure>
</div>
<div class="caption">
<p>from <a href="https://www.nature.com/articles/s41588-019-0483-y">Kelleher et al 2018, <em>Inferring whole-genome histories in large population datasets</em></a>, Nature Genetics</p>
</div>
<!-- Estimated sizes of files required to store the genetic variation data for a
simulated human-like chromosome (100 megabases) for up to 10 billion haploid
(5 billion diploid) samples. Simulations were run for 10 1 up to 10 7 haplotypes
using msprime [Kelleher et al., 2016], and the sizes of the resulting files plotted
(points). -->
</section><section id="history-is-a-sequence-of-trees-1" class="slide level2">
<h2>History is a sequence of trees</h2>
<p>For a set of sampled chromosomes, at each position along the genome there is a genealogical tree that says how they are related.</p>
<figure>
<img data-src="sim_ts.anim.gif" alt="Trees along a chromosome" /><figcaption>Trees along a chromosome</figcaption>
</figure>
</section><section class="slide level2">

<p>A <strong>tree sequence</strong> describes this, er, sequence of trees.</p>
<div class="fragment">
<p><em>Observations:</em></p>
<ol type="1">
<li><p>The <em>pedigree</em> (parental relationships) plus crossover locations would give us the tree sequence for <em>everyone, ever</em>.</p></li>
<li><p>Much less can fully describe the history relevant to a <em>sample</em> of genomes.</p></li>
<li><p>This information is equivalent to the Ancestral Recombination Graph (ARG).</p></li>
</ol>
</div>
</section><section class="slide level2">

<p><a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004842">Kelleher, Etheridge, and McVean</a> introduced the <strong>tree sequence</strong> data structure for a fast coalescent simulator, <a href="https://github.com/jeromekelleher/msprime">msprime</a>.</p>
<ul>
<li><p>stores genealogical <em>and</em> variation data <strong>very</strong> compactly</p></li>
<li><p>efficient algorithms available:</p>
<ul>
<li>subsetting</li>
<li>calculation of allele frequencies in arbitrary cohorts</li>
<li>linkage disequilibrium</li>
<li>log-time haplotype matching</li>
</ul></li>
<li><p>tree-based sequence storage closely related to haplotype-matching compression</p></li>
</ul>
</section><section id="simulated-file-sizes" class="slide level2">
<h2>Simulated file sizes</h2>
<div class="columns">
<div class="column" style="width:50%;">
<figure>
<img data-src="file_size.png" alt="file sizes" /><figcaption>file sizes</figcaption>
</figure>
</div><div class="column" style="width:50%;">
<ul>
<li>HapMap chr1 genetic map (250Mb)</li>
<li>Gutenkunst et al out-of-Africa model (3 pops)</li>
<li>mutation rate <span class="math inline">\(2 \times 10^{-8}\)</span> per gen</li>
<li><p>at <span class="math inline">\(n=10^7\)</span></p>
<ul>
<li>about 17 million variants</li>
<li>VCF size: 318 TiB (250,000<span class="math inline">\(\times\)</span> larger)</li>
</ul></li>
</ul>
</div>
</div>
</section><section id="example-three-samples-two-trees-two-variant-sites" class="slide level2">
<h2>Example: three samples; two trees; two variant sites</h2>
<figure>
<img data-src="example_tree_sequence.png" alt="Example tree sequence" /><figcaption>Example tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<p>Storing a tree sequence in the four tables - <em>nodes</em>, <em>edges</em>, <em>sites</em>, and <em>mutations</em> - is <em>succinct</em> (no redundancy).</p>
<div class="fragment">
<p>These are stored efficiently (hdf5) on disk with a bit more information (e.g., metadata).</p>
</div>
</section><section id="nodes-and-edges" class="slide level2">
<h2>Nodes and edges</h2>
<dl>
<dt>Edges</dt>
<dd><p>Who inherits from who; only <em>necessary</em> for coalescent events.</p>
<p>Records: interval (left, right); parent node; child node.</p>
</dd>
<dt>Nodes</dt>
<dd><p>The ancestors those happen in.</p>
<p>Records: time ago (of birth); ID (implicit).</p>
</dd>
</dl>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.0.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.1.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.2.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.3.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.4.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.5.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="nodes_edges_walkthrough/nodes_edges_walkthrough.6.png" alt="Building a tree sequence" /><figcaption>Building a tree sequence</figcaption>
</figure>
</section><section id="sites-and-mutations" class="slide level2">
<h2>Sites and mutations</h2>
<dl>
<dt>Mutations</dt>
<dd><p>When state changes along the tree.</p>
<p>Records: site it occured at; node it occurred in; derived state.</p>
</dd>
<dt>Sites</dt>
<dd><p>Where mutations fall on the genome.</p>
<p>Records: genomic position; ancestral (root) state; ID (implicit).</p>
</dd>
</dl>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.0.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.1.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.2.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.3.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="sites_muts_walkthrough/sites_muts_walkthrough.4.png" alt="Adding mutations" /><figcaption>Adding mutations</figcaption>
</figure>
<!-- application to fwds sim recording (5min) -->
</section></section>
<section><section id="forwards-time-tree-sequence-recording" class="title-slide slide level1"><h1>Forwards-time tree sequence recording</h1></section><section id="the-main-idea" class="slide level2">
<h2>The main idea</h2>
<p>If we <em>record the tree sequence</em> that relates everyone to everyone else,</p>
<p>after the simulation is over we can put neutral mutations down on the trees.</p>
<div class="fragment">
<p>Since neutral mutations don’t affect demography,</p>
<p>this is <em>equivalent</em> to having kept track of them throughout.</p>
</div>
</section><section class="slide level2">

<p>This means recording the entire genetic history of <strong>everyone</strong> in the population, <strong>ever</strong>.</p>
<div class="fragment">
<p>It is <em>not</em> clear this is a good idea.</p>
</div>
</section><section id="tree-recording-strategy" class="slide level2">
<h2>Tree recording strategy</h2>
<p>Every time an individual is born, we must:</p>
<div>
<ol type="1">
<li class="fragment">add each gamete to the Node Table,</li>
<li class="fragment">add entries to the Edge Table recording which parent each gamete inherited each bit of genome from, and</li>
<li class="fragment">add any new selected mutations to the Mutation Table and (if necessary) their locations to the Site Table.</li>
</ol>
</div>
<div class="fragment">
<div class="columns">
<div class="column" style="width:15%;">
<figure>
<img data-src="finger_right.png" alt="Rightarrow" style="width:100.0%" /><figcaption>Rightarrow</figcaption>
</figure>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:75%;">
<p>This produces <strong>waaaaay</strong> too much data.</p>
</div>
</div>
</div>
</section><section class="slide level2">

<p>We won’t end up needing the <em>entire</em> history of <em>everyone ever</em>,</p>
<div class="fragment">
<p>but we won’t know <em>what</em> we’ll need until later.</p>
</div>
<div class="fragment">
<p>How do we get rid of the extra stuff?</p>
</div>
</section><section id="simplification" class="slide level2">
<h2>Simplification</h2>
<p><em>Question:</em> given a tree sequence containing the history of many individuals, how do we <em>simplify</em> it to only the history of a subset?</p>
</section><section class="slide level2">

<p>Concretely, given an input tree sequence and a subset of its nodes we call the <em>samples</em>, we want a new tree sequence for which:</p>
<ol type="1">
<li><p>All marginal trees match the corresponding subtree in the input tree sequence.</p></li>
<li><p>Every non-sample node in marginal trees has at least two children.</p></li>
<li><p>All nodes and edges are ancestral to at least one sample.</p></li>
<li><p>No adjacent redundant edges (e.g., <span class="math inline">\((\ell, x, p, c) + (x, r, p, c) \rightarrow (\ell, r, p, c)\)</span>).</p></li>
</ol>
</section><section class="slide level2">

<p><em>Answer:</em> to simplify a tree sequence to the history of the <em>samples</em>:</p>
<ol type="1">
<li><p>Paint each <em>sampled</em> chromosome a distinct color.</p></li>
<li><p>Moving back up the tree sequence, copy colors of each chromosome to the parental chromosomes they inherited from.</p></li>
<li><p>If two colors go in the same spot (<em>coalescence</em>), replace with a new color (unique to that ancestor). Output a node for the ancestor and an edge for the coalescence.</p></li>
<li><p>Once all colors have coalesced in a given segment, stop propagating it.</p></li>
</ol>
</section><section id="an-example-simplify-these-to-j-and-k" class="slide level2">
<h2>An example: simplify these to J and K</h2>
<figure>
<img data-src="simplify_walkthrough.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.0.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.1.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.2.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.3.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.4.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.5.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.6.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.7.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.8.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.9.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.10.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.11.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.12.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.13.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.14.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="simplify_walkthrough/simplify_walkthrough.15.png" alt="Simplify example" style="width:100.0%" /><figcaption>Simplify example</figcaption>
</figure>
</section><section id="wright-fisher-n10-before-simplification" class="slide level2">
<h2>Wright-Fisher, N=10: before simplification</h2>
<figure>
<img data-src="sim_wf.anim.gif" alt="Wright-Fisher tree sequence" /><figcaption>Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="wright-fisher-n10-before-simplification-1" class="slide level2">
<h2>Wright-Fisher, N=10: before simplification</h2>
<figure>
<img data-src="sim_wf_unlabeled.anim.gif" alt="Wright-Fisher tree sequence" /><figcaption>Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="and-after-simplification" class="slide level2">
<h2>… and after simplification</h2>
<figure>
<img data-src="sim_wf_simplified.anim.gif" alt="Simplified Wright-Fisher tree sequence" /><figcaption>Simplified Wright-Fisher tree sequence</figcaption>
</figure>
</section><section id="revised-tree-recording-strategy" class="slide level2">
<h2><em>Revised</em> tree recording strategy</h2>
<p>Every time an individual is born, we must:</p>
<ol type="1">
<li>add each gamete to the Node Table,</li>
<li>add entries to the Edge Table recording which parent each gamete inherited each bit of genome from</li>
<li>add any new mutations to the Mutation Table and (if necessary) their locations to the Site Table.</li>
</ol>
<p>… and,</p>
<ol start="4" type="1">
<li>Every so often, <em>simplify</em> the tables so far, retaining the history of the current generation.</li>
</ol>
<!-- 5. display impressive speedups 2min) -->
</section></section>
<section><section id="implementation-and-results" class="title-slide slide level1"><h1>Implementation and results</h1></section><section id="benchmark-implementation" class="slide level2">
<h2>Benchmark implementation</h2>
<ul>
<li><p>Recording, simplifying, and output of tables: <code>C</code> code in <code>msprime</code>.</p></li>
<li><p>Simulation: <a href="https://github.com/molpopgen/fwdpp"><code>fwdpp</code></a>, by Kevin Thornton (in <code>C++</code>) (<a href="https://github.com/molpopgen/fwdpy11_arg_example">code</a>)</p></li>
<li><p>Glue: <a href="https://github.com/pybind/pybind11/"><code>pybind11</code></a> and <a href="http://www.numpy.org/"><code>numpy</code></a></p></li>
<li><p>Machine: Ubuntu / 2x 2.6 GHz Intel E5-2650 CPU</p></li>
</ul>
<div class="fragment">
<p><em>Other implementations:</em></p>
<ul>
<li><a href="https://github.com/ashander/ftprime">pure <code>python</code></a>, interfacing with <a href="https://github.com/BoPeng/simuPOP"><code>simuPOP</code></a></li>
<li><a href="https://github.com/molpopgen/tutorials/blob/cython_cpp_tutorial/notebooks/wfcython.ipynb">cython</a></li>
</ul>
</div>
</section><section id="simulation-parameters" class="slide level2">
<h2>Simulation parameters</h2>
<ol type="1">
<li>Wright-Fisher population of size <span class="math inline">\(N\)</span></li>
<li>simulated for <span class="math inline">\(10N\)</span> genreations</li>
<li>neutral mutation rate <span class="math inline">\(\mu\)</span> equal to recombination rate <span class="math inline">\(r\)</span> per gamete</li>
<li>many, weakly deleterious mutations: rate <span class="math inline">\(\mu/100\)</span> with <span class="math inline">\(s\)</span> exponentially distributed with mean <span class="math inline">\(2.5/N\)</span>.</li>
</ol>
<div class="fragment">
<p><em>Note:</em> if we recorded tree sequences (“pedigree recording”) then the neutral mutation rate was <em>zero</em> but neutral mutations were added <em>afterwards</em>.</p>
</div>
</section><section class="slide level2">

<figure>
<img data-src="rawspeed.png" alt="Total run time per single simulation as a function of region length." style="width:80.0%" /><figcaption>Total run time per single simulation as a function of region length.</figcaption>
</figure>
</section><section class="slide level2">

<figure>
<img data-src="speedup.png" alt="Relative speedup of simulations" style="width:80.0%" /><figcaption>Relative speedup of simulations</figcaption>
</figure>
</section><section id="memory-use" class="slide level2">
<h2>Memory use</h2>
<p>RAM requirements are determined by how often you simplify.</p>
</section></section>
<section><section id="along-the-sequence-of-trees" class="title-slide slide level1"><h1>Along the sequence of trees</h1></section><section id="computation-run-time" class="slide level2">
<h2>Computation run time</h2>
<div class="centered">
<figure>
<img data-src="figs/treestats/benchmarks_without_copy_longer_genome.png" alt="efficiency of treestat computation" style="width:70.0%" /><figcaption>efficiency of treestat computation</figcaption>
</figure>
</div>
<div class="caption">
<p>from <a href="https://www.biorxiv.org/content/10.1101/779132v1?rss=1">Ralph, Thornton and Kelleher 2019, <em>Efficiently summarizing relationships in large samples</em></a>, bioRxiv</p>
</div>
</section><section id="section-3" class="slide level2">
<h2></h2>
<p><strong>What do genotypes tell us about the genealogies?</strong></p>
<div class="caption">
<p>also from <a href="https://www.biorxiv.org/content/10.1101/779132v1?rss=1">Ralph, Thornton and Kelleher 2019, <em>Efficiently summarizing relationships in large samples</em></a>, bioRxiv</p>
</div>
</section><section id="summaries-of-genotypes-and-genealogies" class="slide level2">
<h2>Summaries of genotypes and genealogies</h2>
<div class="columns">
<div class="column" style="width:47%;">
<p><em>Genotypes:</em></p>
<ol type="1">
<li>For each site,</li>
<li>look at who has which alleles,</li>
<li>and add a <em>summary</em> of these values to our running total.</li>
</ol>
<p><em>Example:</em> genetic distance counts how many mutations differ between two sequences.</p>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:47%;">
<!-- nothing til next slide -->
</div>
</div>
</section><section id="summaries-of-genotypes-and-genealogies-1" class="slide level2">
<h2>Summaries of genotypes and genealogies</h2>
<div class="columns">
<div class="column" style="width:47%;">
<p><em>Genotypes:</em></p>
<ol type="1">
<li>For each site,</li>
<li>look at who has inherited which alleles,</li>
<li>and add a <em>summary</em> of these values to the running total.</li>
</ol>
<p><em>Example:</em> sequence divergence counts how many mutations differ between two sequences.</p>
</div><div class="column" style="width:5%;">

</div><div class="column" style="width:47%;">
<p><em>Trees:</em></p>
<ol type="1">
<li>For each branch,</li>
<li>look at who would inherit mutations on that branch,</li>
<li>and add the <em>expected contribution</em> to the running total.</li>
</ol>
<p><em>Example:</em> the mean time to most recent common ancestor between two sequences.</p>
</div>
</div>
</section><section id="section-4" class="slide level2">
<h2></h2>
<figure>
<img data-src="figs/ts_ex/tree_sequence_site_and_branch.png" alt="site and branch stats" /><figcaption>site and branch stats</figcaption>
</figure>
</section><section id="duality" class="slide level2">
<h2>Duality</h2>
<p>Any set of <em>sample weights</em> <span class="math inline">\(w\)</span> and <em>summary function</em> <span class="math inline">\(f\)</span> defines both</p>
<ul>
<li>a statistic of genotypes, <span class="math inline">\(\text{Site}(f,w)\)</span>, and</li>
<li>a statistic of genealogies, <span class="math inline">\(\text{Branch}(f,w)\)</span>.</li>
</ul>
<div class="fragment">
<p>With genealogies <em>fixed</em>, and averaging only over <em>mutations</em> with rate <span class="math inline">\(\mu\)</span>, <span class="math display">\[\begin{equation}
    \text{Branch}(f, w) = \frac{1}{\mu} \E\left[ \text{Site}(f, w) \right] .
\end{equation}\]</span></p>
</div>
<div class="fragment">
<p>Dealing directly with genealogies can <em>remove the layer of noise due to mutation</em>: <span class="math display">\[\begin{equation}
    \frac{1}{\mu^2} \var\left[\text{Site}(f,w)\right]
    =
    \var\left[\text{Branch}(f,w)\right]
    +
    \frac{1}{n}
    \E\left[\text{Branch}(f^2,w)\right]
\end{equation}\]</span></p>
</div>
</section><section id="section-5" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:60%;">
<figure>
<img data-src="figs/treestats/relate_chr20_site_div_branch.1000000.diversity.png" alt="duality in 1000G data" style="width:100.0%" /><figcaption>duality in 1000G data</figcaption>
</figure>
</div><div class="column" style="width:40%;">
<p>Duality, on 1000 Genomes data? Not quite…</p>
<ul>
<li>variation in mutation rate?</li>
<li>biased gene conversion?</li>
<li>selection?</li>
<li>inference artifacts?</li>
</ul>
<div class="caption">
<p><em>Tree sequence from <a href="https://www.nature.com/articles/s41588-019-0484-x">Speidel et al 2019</a>.</em></p>
</div>
</div>
</div>
</section></section>
<section><section id="inferring-the-tree-sequence" class="title-slide slide level1"><h1>Inferring the tree sequence</h1></section><section id="section-6" class="slide level2">
<h2></h2>
<p>What about real data?</p>
</section><section id="inferring-the-ancestry-of-everyone" class="slide level2">
<h2>Inferring the ancestry of everyone</h2>
<p><strong>Left-to-right:</strong></p>
<p>Leo Speidel, Marie Forest, Sinan Shi, and Simon R. Myers. <em>A method for genome-wide genealogy estimation for thousands of samples.</em> Nature Genetics, 51(9):1321–1329, 2019.</p>
<div class="fragment">
<p><strong>Bottom-up:</strong></p>
<p>Jerome Kelleher, Yan Wong, Anthony W. Wohns, Chaimaa Fadil, Patrick K. Albers, and Gil McVean. <em>Inferring whole-genome histories in large population datasets.</em> Nature Genetics, 51(9):1330–1338, 2019.</p>
</div>
</section><section id="challenges-in-inference" class="slide level2">
<h2>Challenges in inference</h2>
<ul>
<li>segmentation of genomes into <em>shared haplotypes</em></li>
<li>hierarchical structuring into a <em>tree sequence</em></li>
<li>calibration of <em>node times</em></li>
</ul>
</section><section id="section-7" class="slide level2">
<h2></h2>
<p>We need <em>metrics</em> for tree sequences.</p>
<div class="fragment">
<ul>
<li class="fragment"><p>Average tree distance along the genome?</p></li>
<li class="fragment"><p>How to quantify haplotype sharing?</p></li>
<li class="fragment"><p>Efficient computation? (should scale to millions of trees with millions of tips)</p></li>
</ul>
</div>
</section></section>
<section><section id="thanks" class="title-slide slide level1"><h1>Thanks</h1></section><section id="acknowledgements" class="slide level2">
<h2>Acknowledgements</h2>
<p>Jerome, Jaime, and Kevin:</p>
<p><a href="http://jeromekelleher.net/pages/about.html"><img data-src="jerome.jpeg" alt="Jerome" style="width:30.0%" /></a>, <a href="http://www.ashander.info/"><img data-src="jaime.jpg" alt="Jaime" style="width:30.0%" /></a> <a href="https://github.com/molpopgen"><img data-src="kevin.jpeg" alt="Kevin" style="width:30.0%" /></a>.</p>
<p>Funding: NSF (PR); Wellcome Trust (JK); NIH (KRT); USF&amp;WS (JDA).</p>
<p>Slides with <a href="http://hakim.se">reveal.js</a> and <a href="https://pandoc.org/">pandoc</a>.</p>
<!-- questions?? -->
</section><section id="section-8" class="slide level2" data-background-image="figs/guillemots_thanks.png" data-background-position="center" data-background-size="50%">
<h2></h2>
</section></section>
    </div>
  </div>

  <script src="reveal.js/lib/js/head.min.js"></script>
  <script src="reveal.js/js/reveal.js"></script>

  <script>

      // Full list of configuration options available at:
      // https://github.com/hakimel/reveal.js#configuration
      Reveal.initialize({
        // Display the page number of the current slide
        slideNumber: true,
        // Push each slide change to the browser history
        history: true,
        // Transition style
        transition: 'none', // none/fade/slide/convex/concave/zoom
        math: {
          mathjax: '/usr/share/javascript/mathjax/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // Optional reveal.js plugins
        dependencies: [
          { src: 'reveal.js/lib/js/classList.js', condition: function() { return !document.body.classList; } },
          { src: 'reveal.js/plugin/zoom-js/zoom.js', async: true },
          { src: 'reveal.js/plugin/math/math.js', async: true },
          { src: 'reveal.js/plugin/notes/notes.js', async: true }
        ]
      });
    </script>
    </body>
</html>
